File: bindings/python/samples/stream_packet_recv_sample.py
import socket
import struct

import zlink
from sample_support import tcp_endpoint, wait_socket_monitor_event


def main():
# --8<-- [start:doc]
    port, endpoint = tcp_endpoint()

    with zlink.create_context() as ctx:
        with zlink.create_stream_socket(ctx) as server:
            server.stream_options.recv_mode = zlink.StreamRecvMode.PACKET
            with server.monitor_open(zlink.MonitorEventMask.ACCEPTED) as server_monitor:
                server.bind(endpoint)
                with socket.create_connection(("127.0.0.1", port), timeout=3.0) as client:
                    wait_socket_monitor_event(server_monitor, zlink.MonitorEventMask.ACCEPTED)
                    client.sendall(
                        struct.pack("!HI", 0, len(b"hello-stream")) + b"hello-stream"
                    )
                    packet = zlink.StreamPacket()
                    if not server.recv_packet_into(packet):
                        raise AssertionError("expected stream packet")
                    with packet:
                        if packet.header.to_bytes() != b"":
                            raise AssertionError("unexpected stream packet header")
                        if packet.body.to_bytes() != b"hello-stream":
                            raise AssertionError("unexpected stream packet body")
                        if not packet.routing_id:
                            raise AssertionError("stream packet expected a routing id")
                        server.send(packet.routing_id).message(b"hello-stream").submit_sync()
                    reply = client.recv(64)
                    if reply != b"hello-stream":
                        raise AssertionError(f"unexpected stream packet reply: {reply!r}")
            print('[stream/packet-recv] send: "hello-stream" -> recv: "hello-stream"')
# --8<-- [end:doc]


if __name__ == "__main__":
    main()

File: bindings/rust/samples/stream_packet_recv_sample.rs
//! STREAM packet sample – demonstrates pull-based framed receive.
//! The STREAM socket binds as a server; a raw TCP client connects inward.

#[path = "sample_support.rs"]
mod sample_support;

use std::io::Write;
use std::net::TcpStream;
use zlink::{Context, RecvFlags, SocketMonitor, StreamPacket, StreamRecvMode};

fn write_stream_packet(stream: &mut TcpStream, body: &[u8]) {
    let mut frame = Vec::with_capacity(6 + body.len());
    frame.extend_from_slice(&0u16.to_be_bytes());
    frame.extend_from_slice(&(body.len() as u32).to_be_bytes());
    frame.extend_from_slice(body);
    stream.write_all(&frame).expect("tcp write failed");
    stream.flush().expect("tcp flush failed");
}

fn main() {
    // --8<-- [start:doc]
    let ctx = Context::new().expect("context creation failed");

    let stream = ctx.stream_socket().expect("stream socket failed");
    stream
        .stream_options()
        .set_recv_mode(StreamRecvMode::Packet)
        .expect("set packet mode failed");
    let endpoint = sample_support::tcp_endpoint();
    stream.bind(&endpoint).expect("bind failed");
    let stream_mon = SocketMonitor::open(&stream).expect("stream monitor open failed");

    let tcp_addr = endpoint.strip_prefix("tcp://").unwrap();
    let mut tcp_client = TcpStream::connect(tcp_addr).expect("tcp connect failed");
    tcp_client.set_nodelay(true).expect("set_nodelay failed");
    sample_support::wait_stream_connected(&stream_mon);
    drop(stream_mon);

    write_stream_packet(&mut tcp_client, b"hello-stream");
    let mut packet = StreamPacket::empty();
    assert!(
        stream
            .recv_packet(&mut packet, RecvFlags::NONE)
            .expect("packet receive failed")
    );
    assert!(
        !packet
            .routing_id()
            .expect("packet routing id")
            .as_bytes()
            .is_empty()
    );
    assert!(
        packet
            .header()
            .expect("packet header")
            .as_bytes()
            .is_empty()
    );
    let payload = packet.body().expect("packet body").as_bytes();
    assert_eq!(payload, b"hello-stream");
    let recv_str = std::str::from_utf8(payload).unwrap();
    println!(
        "[stream/packet-recv] send: \"hello-stream\" → recv: \"{}\"",
        recv_str
    );
    packet.close().expect("packet close failed");
    // --8<-- [end:doc]
}
